### DISCUSSION TITLE
Demo Source Code

### DISCUSSION THREAD

**Discussion by mosesman831** [vpnhood/VpnHood#122] [Q&A] [answered]
Hello guys! So I am trying to make a custom VPN client but I am kind of confused at the docs. Do anyone have a demo one (.sln with some source files, the solution can be built or not built) Please help cause I want to make this for so long using OpenVPN or WireGuard but I think this is it, this will save months of researching. Thank you!

Btw Does the user see any of the server details? Trying to make a professional VPN for a school project. Thank you!

**Comment by bobvhood** [comment, marked as answer]
Hi,
VpnHood client can just connect to VpnHood server, you can not connect to OpenVPN server or Wireguard.
Here is a client sample, let us know if it doesn't work.
https://github.com/vpnhood/VpnHood/tree/main/Samples

> Btw Does the user see any of the server details? 

No, they will not unless you share an accessKey.

**Reply by mosesman831** [reply]
Is there one with GUI?

**Reply by bobvhood** [reply]
No, the sample does not have GUI because it doesn't require to show how it works. You can create your own UI in by different technologies. 
We used HTML and Vue. you can see find it here: https://github.com/vpnhood/VpnHood.AccessServer.UI

I recommend you create your own UI.

**Comment by mosesman831** [comment]
pls check it out. 
![image](https://user-images.githubusercontent.com/92276967/176970069-50718544-7f4f-4594-9790-446a00d3e498.png)
Error

**Reply by bobvhood** [reply]
Hi,
1) Is it trying to connect to our public servers? If yes, make sure you can connect to our public servers via our compiled app.
2) Try to get the latest version from our development branch and let me if it doesn't work.

**Reply by mosesman831** [reply]
Im using the sample version...

**Reply by bobvhood** [reply]
![image](https://user-images.githubusercontent.com/92276967/176992364-63bf1401-8d63-4c9d-9a20-ecf7e09bd5c6.png)
I couldn't find this message in the source code, where did you download it?

**Reply by mosesman831** [reply]
https://github.com/vpnhood/VpnHood/tree/main/Samples
Used my own server address
accesskey
photo:
https://ibb.co/TWy1BNh

**Comment by mosesman831** [comment]
It’s the sample file u sent me

**Reply by bobvhood** [reply]
use this one:
https://github.com/vpnhood/VpnHood/tree/development/Samples

**Reply by mosesman831** [reply]
It seems like my server access file is wrong, but also can you add max client counts?

**Reply by mosesman831** [reply]
Compared to the public VPN access code, mine is much shorter and has a = in the end. 
Sent censored access key to your email.

**Reply by bobvhood** [reply]
try to add your accessKey to VpnHoodClient official app., then you should see your server publicIP in the app.
If you can't connect it via VpnHood client app, it means your server doesn't listen. check your ports, and firewall.

**Reply by mosesman831** [reply]
What port and address should I unblock?

**Comment by bobvhood** [comment]
Sorry, I miss the conversation. Could you connect to your server?

**Reply by mosesman831** [reply]
I don't think it can connect because of some OCI firewall rules. Do you have some specific ports or other stuff you will need to unblock to use it?

**Reply by bobvhood** [reply]
When you run the VpnHoodServer it shows the listening port. You should make sure they are open. Also, you can find it in your accessKey.

**Reply by mosesman831** [reply]
Does it work with Access server

**Reply by mosesman831** [reply]
Info 
| 21:43:34.5003 |  | AppVersion: 2.4.303.0

Info 
| 21:43:34.5009 |  | Time: 2022-07-05 20:43:34Z

Info 
| 21:43:34.5014 |  | OS: Microsoft Windows NT 10.0.19044.0, 64-bit

Info 
| 21:43:34.5020 |  | UserAgent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/103.0.5060.66 Safari/537.36 Edg/103.0.1264.44

Info 
| 21:43:35.9502 |  | Country: United Kingdom

Info 
| 21:43:35.9508 |  | TokenId: **-b716-e154cd377434, SupportId: **02

Trac 
| 21:43:35.9521 |  | Checking the Internet connection...

Info 
| 21:43:35.9526 |  | HttpTest: Started, Uri: https://www.google.com/, Timeout: 10000...

Info 
| 21:43:35.9532 |  | HttpTest: Started, Uri: https://www.quad9.net/, Timeout: 10000...

Info 
| 21:43:35.9538 |  | HttpTest: Started, Uri: https://www.microsoft.com/, Timeout: 10000...

Info 
| 21:43:36.0379 |  | HttpTest: Succeeded, Started, Uri: https://www.microsoft.com/.

Trac 
| 21:43:36.0389 |  | Checking the VpnServer ping...

Info 
| 21:43:36.0394 |  | PingTest: Started, RemoteAddress: 140.238.68.248, Timeout: 10000...

Info 
| 21:43:36.0897 |  | HttpTest: Succeeded, Started, Uri: https://www.google.com/.

Info 
| 21:43:36.8517 |  | HttpTest: Succeeded, Started, Uri: https://www.quad9.net/.

Erro 
| 21:43:45.8269 |  | PingTest: Failed! RemoteAddress: 140.*.*.248, Message: Status: TimedOut.

Warn 
| 21:43:45.8277 |  | Could not ping server! EndPont: 140.*.*.248:443, Error: Status: TimedOut

Info => Client
| 21:43:45.8285 |  | UseUdpChannel: True, MinWorkerThreads: 16, CompletionPortThreads: 16

Info => Client
| 21:43:45.8288 |  | Client Version: 2.4.303.0, ClientId: **-a675-c73894fefaea

Info => Client
| 21:43:45.8294 |  | Client is Connecting

Trac => Client
| 21:43:45.8298 | Session | Connecting to Server: 140.*.*.248:443...

Erro => Client
| 21:44:06.8843 |  | Disposing. Error! System.Net.Sockets.SocketException (10060): A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond.
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.ThrowException(SocketError error, CancellationToken cancellationToken)
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.System.Threading.Tasks.Sources.IValueTaskSource.GetResult(Int16 token)
   at System.Threading.Tasks.ValueTask.ValueTaskSourceAsTask.<>c.<.cctor>b__4_0(Object state)
--- End of stack trace from previous location ---
   at System.Net.Sockets.TcpClient.CompleteConnectAsync(Task task)
   at VpnHood.Common.Util.RunTask(Task task, TimeSpan timeout, CancellationToken cancellationToken) in C:\Users\Robert\Source\Repos\Vh\VpnHood\VpnHood.Common\Util.cs:line 109
   at VpnHood.Client.VpnHoodClient.GetSslConnectionToServer(EventId eventId, CancellationToken cancellationToken) in C:\Users\Robert\Source\Repos\Vh\VpnHood\VpnHood.Client\VpnHoodClient.cs:line 605
   at VpnHood.Client.VpnHoodClient.ConnectInternal(CancellationToken cancellationToken, Boolean redirecting) in C:\Users\Robert\Source\Repos\Vh\VpnHood\VpnHood.Client\VpnHoodClient.cs:line 666
   at VpnHood.Client.VpnHoodClient.Connect() in C:\Users\Robert\Source\Repos\Vh\VpnHood\VpnHood.Client\VpnHoodClient.cs:line 210

Trac => Client
| 21:44:06.8849 |  | Disconnecting...

Info => Client
| 21:44:06.8854 |  | Client is Disconnecting

Trac => Client
| 21:44:06.8858 |  | Shutting down...

Trac => Client
| 21:44:06.8864 |  | Disposing TcpProxyHost...

Trac => Client
| 21:44:06.8870 |  | Disposing Tunnel...

Trac => Client
| 21:44:06.8876 |  | Disposing ProxyManager...

Trac => Client
| 21:44:06.8882 |  | Disposing Nat...

Info => Client
| 21:44:06.8888 |  | Client is Disposed

Info => Client
| 21:44:06.8893 |  | Bye Bye!

Erro 
| 21:44:06.8902 |  | A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond.

Diagnose report

**Reply by mosesman831** [reply]
Should I show you my OCI NSG?

**Comment by mosesman831** [comment]
@bobvhood Should I show you my OCI NSG?

**Reply by bobvhood** [reply]
> OCI NSG

What is OCI NSG? I am not familiar with that.

**Reply by mosesman831** [reply]
Oracle Cloud infrastructure– Network security group

**Reply by bobvhood** [reply]
You already sent me your client log. It just shows that no listener is running on your server at that port and ip. Can you send me your VpnHood server log (not the client)?

**Reply by mosesman831** [reply]
https://ibb.co/42bVx7G

scrnsht of Access Server

**Reply by bobvhood** [reply]
Would you please copy/paste your image directly into this discussion?

**Reply by mosesman831** [reply]
<img width="956" alt="Screenshot 2022-07-06 202355" src="https://user-images.githubusercontent.com/56060463/177639782-f434428c-0760-45a8-8c74-0a19d37b956d.png">
oh srry it worked for once!

**Reply by bobvhood** [reply]
Ok. Actually, it is just our access server page without any information :)
However, is your public ip is same as your private IP? Are you sure TCP 443 port is open on your network firewall?

**Reply by mosesman831** [reply]
@bobvhood they are different. I also tried unblocking all tcp ports but it still doesn't work...

**Reply by bobvhood** [reply]
Please go to your server (your machine), find the log and email it to support@vpnhood.com
it should be in /opt/VpnHoodServer

**Reply by mosesman831** [reply]
GNU nano 6.2                                              log.txt                                                       
2022-07-05 20:39:42.1206 | Info |
 Initializing ResetAccessServer. BaseUrl: https://access.vpnhood.com/api/agent/

2022-07-05 20:39:42.8108 | Info |
 VpnHoodServer, Version=2.4.303.0, Culture=neutral, PublicKeyToken=null

2022-07-05 20:39:42.8108 | Info |
 OS: Unix 5.15.0.1011, 64-bit
Linux version 5.15.0-1011-oracle (buildd@lcy02-amd64-066) (gcc (Ubuntu 11.2.0-19ubuntu1) 11.2.0, GNU ld (GNU Binutils for>

2022-07-05 20:39:43.8205 | Info | Server
 MinWorkerThreads: 2, MinCompletionPortThreads: 60, MaxWorkerThreads: 32767, MaxCompletionPortThreads: 32767,

2022-07-05 20:39:43.8258 | Info | Server
 Configuring by the Access Server...

2022-07-05 20:39:45.5246 | Info | Server
 ServerConfig: {"TrackingOptions":{"LogLocalPort":false,"LogClientIp":false},"SessionOptions":{"Timeout":"00:10:00","UdpT>

2022-07-05 20:39:45.5269 | Info | Server
 Starting TcpHost...

2022-07-05 20:39:45.5269 | Info | Server
 Start listening on 10.*.*.57:443

2022-07-05 20:39:45.6074 | Info | Server
 Set UpdateStatusInterval to 60 seconds.

2022-07-05 20:39:45.6074 | Info | Server
 Server is ready!

**Reply by mosesman831** [reply]
@bobvhood there isn't any error messages, will it just be a firewall issue?

**Reply by trudyhood** [reply]
Hi All,

I think the network firewall of your server causes the problem. Try to disable your ubuntu firewall and make sure to open the 443 port on your network enterprise firewall. 
Also, you have something like Amazon ElasticIP (I do not know about Oracle Cloud infrastructure). Make sure your public IP in the token is mapped to your server's local IP. Your server is listening to 10...57:443 (perhaps local IP), and your public IP is different, as you said. So make sure your network infrastructure maps it correctly. If you find it is different, try to change your public IP in our access server and get a new accessKey.

**Reply by mosesman831** [reply]
Yeah my public ip is my public ip on Oracle cloud (local IP)
tried everything but still doesn't work

**Reply by trudyhood** [reply]
Can you run any web server on your server and connect to it? Can you ping your server?

**Reply by mosesman831** [reply]
@bobvhood Request timed out
even i have flushed all iptables and setup Network security group...

**Reply by bobvhood** [reply]
Request timed out! Who gives you that error and how do you get it?

**Reply by mosesman831** [reply]
CMD
ping [server address]

**Reply by bobvhood** [reply]
You need to get help from the network administrator. VpnHood requires basic network functionality.
